x/valset-pref/msg_server_test.go

```diff
@@ -347,75 +347,77 @@ func (suite *KeeperTestSuite) TestRedelegateValidatorSet() {
 					Weight:         sdk.NewDecWithPrec(2, 1),
 				},
 			},
-			coinToStake:     sdk.NewCoin(sdk.DefaultBondDenom, sdk.NewInt(20_000_000)),
-			expectedShares:  []sdk.Dec{sdk.NewDec(10_000_000), sdk.NewDec(6_000_000), sdk.NewDec(4_000_000)},
-			expectPass:      false,
-			delegationExist: true,
+			coinToStake:           sdk.NewCoin(sdk.DefaultBondDenom, sdk.NewInt(20_000_000)),
+			expectedShares:        []sdk.Dec{sdk.NewDec(10_000_000), sdk.NewDec(6_000_000), sdk.NewDec(4_000_000)},
+			valSetDelegationExist: true,
+			expectPass:            false, // cannot redelegate to same validator set
```

Comment:
Why is this the design? If I delegate to A (0.2) B (0.5) C (0.3) and want to redelegate to balance to A (0.3) B (0.5) C (0.2), why doesn't this do a single redelegate message from C to A?

Reply:
So in the sdk it checks if sourceAddr == destAddr ([code](https://github.com/cosmos/cosmos-sdk/blob/main/x/staking/keeper/delegation.go#L899))

But this is causing because of our redelegation algorithm, for ex: 
```
Existing Set {A: 0.2, B: 0.5, C: 0.3}, NewSet {A: 0.3, B: 0.5, C: 0.2} 
Delegate 20osmo 
ExistingSet{A: 4, B: 10, C: 6}, NewSet{A: 6, B: 10, C: 4}, DiffSet: {A:4, B:10, C:6, A: -6, B: -10, C: -4} 

Perform Redelegation
A --redelegate--> B (4osmo) 
B --redelegate--> B (6osmo) (ERROR: Self delegation) 

** Suppose we force the source and destination address to be different** 
Perform Redelegation
A --redelegate--> B (4osmo) 
B --redelegate--> A (6osmo) (ERROR: ValA is already in redelegation) 
```

To fix this the algorithm should be aware of the val addresses it's using and also if redelegation is possible from that address which could make things a lil complicated

Reply:
I thought about this a lot and there's still issues even if we make sourceValAddr != targetValAddr

something like this for instance; 

> "If I delegate to A (0.2) B (0.5) C (0.3) and want to redelegate to balance to A (0.3) B (0.5) C (0.2), why doesn't this do a single redelegate message from C to A?" 

Redelegation has to be done from C --> A, and then A --> C. But A --> C isn't possible because "Redelegation to ValidatorA is in progress and must complete before we perform another redelegation"